components/brave_wallet/common/fil_address.cc

```diff
@@ -0,0 +1,180 @@
+    size_t length) {
+  blake2b_state blakeState;
+  if (blake2b_init(&blakeState, length) != 0) {
+    VLOG(0) << __func__ << ": blake2b_init failed";
+    return absl::nullopt;
+  }
+  if (blake2b_update(&blakeState, payload.data(), payload.size()) != 0) {
+    VLOG(0) << __func__ << ": blake2b_update failed";
+    return absl::nullopt;
+  }
+  std::vector<uint8_t> result(length, 0);
+  if (blake2b_final(&blakeState, result.data(), length) != 0) {
+    VLOG(0) << __func__ << ": blake2b_final failed";
+    return absl::nullopt;
+  }
+  return result;
+}
+
+bool IsValidNetwork(const std::string& network) {
+  return network == mojom::kFilecoinTestnet ||
+         network == mojom::kFilecoinMainnet;
+}
+
+absl::optional<mojom::FilecoinAddressProtocol> ToProtocol(char input) {
+  if ((input - '0') == static_cast<int>(mojom::FilecoinAddressProtocol::BLS))
+    return mojom::FilecoinAddressProtocol::BLS;
+  if ((input - '0') ==
+      static_cast<int>(mojom::FilecoinAddressProtocol::SECP256K1))
+    return mojom::FilecoinAddressProtocol::SECP256K1;
+  return absl::nullopt;
+}
+
+}  // namespace
+
+FilAddress::FilAddress(const std::vector<uint8_t>& bytes,
+                       mojom::FilecoinAddressProtocol protocol,
+                       const std::string& network)
+    : protocol_(protocol), network_(network), bytes_(bytes) {
+  DCHECK(IsValidNetwork(network));
+}
+FilAddress::FilAddress() = default;
+FilAddress::FilAddress(const FilAddress& other) = default;
+FilAddress::~FilAddress() = default;
+
+bool FilAddress::IsEqual(const FilAddress& other) const {
+  return bytes_.size() == other.bytes_.size() &&
+         std::equal(bytes_.begin(), bytes_.end(), other.bytes_.begin()) &&
+         protocol_ == other.protocol_ && network_ == other.network_;
+}
+
+bool FilAddress::operator==(const FilAddress& other) const {
+  return IsEqual(other);
+}
+
+bool FilAddress::operator!=(const FilAddress& other) const {
+  return !IsEqual(other);
+}
+
+// Decodes Filecoin BLS/SECP256K addresses within rules
+// https://spec.filecoin.io/appendix/address/#section-appendix.address.string
+// |------------|----------|---------|----------|
+// |  network   | protocol | payload | checksum |
+// |------------|----------|---------|----------|
+// | 'f' or 't' |  1 byte  | n bytes | 4 bytes  |
+// static
+FilAddress FilAddress::FromAddress(const std::string& address) {
+  if (address.size() != kAddressSizeBLS &&
+      address.size() != kAddressSizeSecp256K)
+    return FilAddress();
+
+  auto protocol = ToProtocol(address[1]);
+  if (!protocol)
+    return FilAddress();
+
+  std::string network{address[0]};
+  if (!IsValidNetwork(network))
+    return FilAddress();
+
+  std::string payload{
+      base32::Base32Decode(base::ToUpperASCII(address.substr(2)))};
+  if (payload.empty())
+    return FilAddress();
+
+  std::string key{payload.substr(0, payload.size() - kChecksumSize)};
+  std::vector<uint8_t> public_key(key.begin(), key.end());
```

Comment:
The thing you're passing in is not public_key but payload.

Reply:
done